uefi/src/system.rs

```diff
@@ -0,0 +1,94 @@
+//! TODO
+
+// TODO
+#![allow(clippy::missing_safety_doc)]
+
+use crate::proto::console::text;
+use crate::table::{cfg, Boot, Revision, SystemTable};
+use crate::CStr16;
+use core::ptr::{self, NonNull};
+use core::slice;
+use core::sync::atomic::{AtomicPtr, Ordering};
+
+static SYSTEM_TABLE: AtomicPtr<uefi_raw::table::system::SystemTable> =
+    AtomicPtr::new(ptr::null_mut());
+
+/// TODO
+pub unsafe fn set_system_table(system_table: *mut uefi_raw::table::system::SystemTable) {
+    SYSTEM_TABLE.store(system_table, Ordering::Release);
+}
+
+/// TODO
+pub(crate) fn system_table_maybe_null() -> *mut uefi_raw::table::system::SystemTable {
+    SYSTEM_TABLE.load(Ordering::Acquire)
+}
+
+/// TODO
+pub(crate) fn system_table() -> NonNull<uefi_raw::table::system::SystemTable> {
+    let st = system_table_maybe_null();
+    NonNull::new(st).expect("set_system_table has not been called")
+}
+
+/// TODO
+#[must_use]
+pub fn system_table_boot() -> SystemTable<Boot> {
+    unsafe { SystemTable::<Boot>::from_ptr(system_table().as_ptr().cast()) }.unwrap()
+}
+
+// TODO: is static lifetime OK for these returned references?
+
+/// Return the firmware vendor string
+#[must_use]
+pub fn firmware_vendor() -> &'static CStr16 {
+    unsafe { CStr16::from_ptr(system_table().as_mut().firmware_vendor.cast()) }
```

Comment:
Why did you decide to use functions here and not a new type having associated functions/methods?

Reply:
I left some notes here: https://github.com/rust-osdev/uefi-rs/issues/893#issuecomment-1656811575

Since there's no longer any struct or lifetime involved, I think these functions are similar to https://doc.rust-lang.org/std/env/#functions; no point attaching them to an empty struct.